cpp/include/raft/spatial/knn/detail/ivf_pq_build.cuh

```diff
@@ -0,0 +1,1028 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../ivf_pq_types.hpp"
+#include "ann_kmeans_balanced.cuh"
+#include "ann_utils.cuh"
+
+#include <raft/core/handle.hpp>
+#include <raft/core/logger.hpp>
+#include <raft/core/mdarray.hpp>
+#include <raft/core/nvtx.hpp>
+#include <raft/cuda_utils.cuh>
+#include <raft/distance/distance_type.hpp>
+#include <raft/linalg/gemm.cuh>
+#include <raft/matrix/matrix.cuh>
+#include <raft/pow2_utils.cuh>
+#include <raft/stats/histogram.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/managed_memory_resource.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+#include <rmm/mr/device/pool_memory_resource.hpp>
+
+namespace raft::spatial::knn::ivf_pq::detail {
+
+using namespace raft::spatial::knn::detail;  // NOLINT
+
+namespace {
+HDI void ivfpq_encode_core(
+  uint32_t n_rows, uint32_t pq_dim, uint32_t pq_bits, const uint32_t* label, uint8_t* output)
+{
+  for (uint32_t j = 0; j < pq_dim; j++, label += n_rows) {
+    uint8_t code = *label;
+    if (pq_bits == 8) {
+      uint8_t* out = output + j;
+      out[0]       = code;
+    } else if (pq_bits == 7) {
+      uint8_t* out = output + 7 * (j / 8);
+      if (j % 8 == 0) {
+        out[0] = code;
+      } else if (j % 8 == 1) {
+        out[0] |= code << 7;
+        out[1] = code >> 1;
+      } else if (j % 8 == 2) {
+        out[1] |= code << 6;
+        out[2] = code >> 2;
+      } else if (j % 8 == 3) {
+        out[2] |= code << 5;
+        out[3] = code >> 3;
+      } else if (j % 8 == 4) {
+        out[3] |= code << 4;
+        out[4] = code >> 4;
+      } else if (j % 8 == 5) {
+        out[4] |= code << 3;
+        out[5] = code >> 5;
+      } else if (j % 8 == 6) {
+        out[5] |= code << 2;
+        out[6] = code >> 6;
+      } else if (j % 8 == 7) {
+        out[6] |= code << 1;
+      }
```

Comment:
Nitpick (I saw this code while looking for something else in the PR). I'm curious if there's any reason not to do:

```cpp
uint32_t jmod8 = j % 8;
if(jmod8 > 0) out[jmod8 - 1] = code << (8 - jmod8);
if(jmod8 < 7) out[jmod8] = code << jmod8;
```

(and you can probably do similar arithmetic in the code below)

Reply:
Thanks, yes, I've ported this code recently due to a bug and was afraid to touch it just yet :) I'm actually thinking whether I can define these as bitfields+templates, but your suggestion looks very nice too.